Source/WebCore/css/color/CSSUnresolvedColorMix.cpp

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "CSSUnresolvedColorMix.h"
+
+#include "ColorFromPrimitiveValue.h"
+#include "ColorSerialization.h"
+#include "StyleBuilderState.h"
+
+namespace WebCore {
+
+static void serializationForCSS(StringBuilder& builder, const CSSUnresolvedColorMix::Component& component)
+{
+    builder.append(component.color->customCSSText());
+    if (component.percentage)
+        builder.append(' ', component.percentage->customCSSText());
+}
+
+String serializationForCSS(const CSSUnresolvedColorMix& unresolved)
```

Comment:
Maybe this should be a pair of functions. Always have the one that takes the `SteingBuilder` and the one without the `StringBuilder` that returns the `String`.

Reply:
That seems like the right pattern. Would be interesting to see if we could make one generate efficiently from the other (or both from a more generic third) at some point as well.